apm-agent-core/src/main/java/co/elastic/apm/agent/impl/baggage/Baggage.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.impl.baggage;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class Baggage implements co.elastic.apm.agent.tracer.Baggage {
+
+    public static final Baggage EMPTY = new Baggage(Collections.<String, String>emptyMap(), Collections.<String, String>emptyMap());
+
+    private final Map<String, String> baggage;
+
+    /**
+     * W3C headers allow to add metadata key-value pairs to baggage.
+     * We do currently not decode these key-value pairs, but propagate them in encoded form
+     * to ensure they are not lost.
+     * <p>
+     * Keys of this map are guaranteed to also be present as keys in #baggage
+     */
+    private final Map<String, String> baggageMetadata;
+
+    /**
+     * Baggage instances are immutable, therefore we can safely cache the serialized form.
+     */
+    private volatile String cachedSerializedW3CHeader = null;
+
+    private Baggage(Map<String, String> baggage, Map<String, String> baggageMetadata) {
+        this.baggage = baggage;
+        this.baggageMetadata = baggageMetadata;
+    }
+
+    public Set<String> keys() {
+        return baggage.keySet();
+    }
+
+    public @Nullable String get(String key) {
+        return baggage.get(key);
+    }
+
+    @Nullable
+    public String getMetadata(String key) {
+        return baggageMetadata.get(key);
+    }
+
+    public boolean isEmpty() {
+        return baggage.isEmpty();
+    }
+
+    public static Builder builder() {
+        return EMPTY.toBuilder();
```

Comment:
I don't understand why you would want a builder that holds immutable empty maps so cannot be changed?

Reply:
I see, you're replacing them if they are from EMPTY. It's not intuitive

Reply:
It does minimize the object creation, I see that's why. Okay, ignore these comments